### PR TITLE
Cloud deployment for history service and enhancement to matching lobby

### LINKED ISSUE #100: Research on AWS Tools
- [x] spend time research on how to push images into ECR
- [ ] API gateway (ready) & authentication (HOLD)
- [ ] try hosting database (RDS) in AWS (HOLD, RDS might be costly)
- [x] Amplify, ECS

### LINKED ISSUE #111: Mapping Services to AWS Api Gateway
### Summary
Each service will need to prepend staging and domain to its endpoint paths.
Example: `/api/questions` -> `/question/api/questions`

### Background
Currently, AWS API gateway's endpoints comprise 3 components: `public-addr`, ~~`staging`~~ and `routes`. 
- public-addr is the publicly accessible URL for our frontends to consume
- ~~staging is used to identify the deployment environment for our APIs, we have `development` and `production` currently.~~
- routes basically define the resource paths of our micro-services. Ideally, we shouldn't need to change our services and hopes API gateway would help us map the public path to private paths which our services then could understand. **that's however not the case.....**

### ~~Problem 1~~ (Switching to multi-stack approach (2 different api gateway url for dev and prod respectively), dropping $stage in endpoints)
~~Instead of doing some conversion, the API gateway simply uses the routes defined as an access control list and simply throws the entire path to our containers directly.~~
Example:
1. ~~public request `id.public-addr.com/development/questions`,~~
2. ~~gateway checks for `/questions/` defined in routes for permission,~~
3. ~~gateway then throws `/development/questions` to question-service directly...~~
4. ~~question-service 404 lor, since it does not understand  `/development/questions`.~~

### Problem 2
Since we might have top-level resources like 'health' in our containers, mapping `/health` directly behind `/stage/health` would cause conflicts in gateway routes.

### Resolution
1. ~~So now the services need to handle staging values in the request path. we can simply replace the `/api` path with our staging value like this:~~

2. To resolve conflicts for our top-level paths, we'll need to introduce a `domain` value in our paths.
```ts
app.use(`/<DOMAIN>/api`, authMiddleware, router);
```

The routes in API gateway will be like this:
image

@tryyang2001  @weidak  @tlyi  @xingjie99 Take note of the changes 🫠

### Tracker
Backend:
- [x] question-service
- [x] matching-service
- [x] user-service
- [x] collaboration-service
- [x] auth-service
- [ ] history-service

Frontend:
- [x] endpoint.ts
- [x] question_api_warpper
- [x] matching_api_warpper
- [x] auth_api_warpper
- [x] collaboration_api_warpper
- [ ] history_api_warpper
- [x] user_api_warpper

### LINKED ISSUE #134: Deploy history service to cloud
- containerize history as a docker container
- expose all env vars needed in the Dockerfile (currently we only need the DATABASE_URL)
- deploy history service to aws and successfully integrate it with existing aws services

### LINKED ISSUE #136: Deploy history service to cloud
containerize history as a docker container
expose all env vars needed in the Dockerfile (currently we only need the DATABASE_URL)
deploy history service to aws and successfully integrate it with existing aws services

### PR DESCRIPTION
**Pull request by c0j0s**
# Pull Request

## Description

Finally, our services are fully on cloud. 

## Related Issue(s)

Close #134 
Close #136 
Close #100 

## Changes Made

- add dockerfile for history service.
- add history service task definition to aws stack template.
- enable ecr deployment for history service.
- add support for language selector in matching lobby.

## Screenshots (if applicable)

image

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

### CODE REVIEW

**Comment by weidak** [comment]
image

I kept getting stuck here.. not sure if something went wrong with the switching of states? I have shut down and re-ran the applications too. Btw, for some reason question detail/history detail does not expand to full screen now on your branch (and on aws):

image

**Comment by weidak** [comment]
Switched back to my branch, and the question detail seems fine:

image

**Comment by weidak** [comment]
Other than that, thank you for deploying to AWS! Seems like its working well 👍🏻

File: .aws/peerprep-stack.yaml
Comment (weidak) [inline .aws/peerprep-stack.yaml:1]: Oops! sorry i forgot about this.. thank you!

**Comment by c0j0s** [comment]
> I kept getting stuck here.. not sure if something went wrong with the switching of states? I have shut down and re-ran the applications too. Btw, for some reason question detail/history detail does not expand to full screen now on your branch (and on aws):
> 
1. should be fixed.
2. not sure what the root cause is, but the latest deployment seems fine.

**Review by weidak — APPROVED** [review]

**Pull request merged**